waiter/src/waiter/scheduler.clj

```diff
@@ -842,8 +842,54 @@
                                  ([service-id] (retrieve-syncer-state @syncer-state-atom service-id)))}))
 
 ;;
-;; Support for tracking killed instances
+;; Support for tracking killed instance candidates
 ;;
+(defn- cleanup-kill-candidate
+  "Cleans up state for a kill candidate.
+   If candidate was not killed but still failed, we trigger the failed instance callback."
+  [kill-candidates-state instance-id]
+  (let [{:keys [track-failed-instance tracked-states]} (get kill-candidates-state instance-id)]
+    (when (and track-failed-instance
+               (contains? tracked-states :failed)
+               (not (contains? tracked-states :killed)))
+      (log/info "tracking kill candidate as failed instance" {:id instance-id})
+      (track-failed-instance))
+    (dissoc kill-candidates-state instance-id)))
+
+(let [kill-candidates-state-atom (atom {})]
+
+  (defn track-kill-candidate!
+    "Tracks which instances are being used as kill candidates for the provided duration."
+    [instance-id reason duration-ms]
+    (log/info "tracking instance as kill candidate" {:duration-sm duration-ms :id instance-id :reason reason})
+    (swap! kill-candidates-state-atom update instance-id update :tracked-states set/union #{reason})
```

Comment:
```suggestion
    (swap! kill-candidates-state-atom update-in [instance-id :tracked-states] conj reason)
```

Reply:
`reason` is initially `nil`, `conj` doesn't get me a set :(

Reply:
😢

Reply:
Just the `update-in` change then...

```suggestion
    (swap! kill-candidates-state-atom update-in [instance-id :tracked-states] set/union #{reason})
```

Reply:
Committed suggestion